### PR TITLE
MPS: MPSGraph created with wrong axes when dim=[] specified

### LINKED ISSUE #97231: Inconsistent behavior and crash in torch.norm MPS implementation
### 🐛 Describe the bug

I've found some issues in the [torch.norm](https://pytorch.org/docs/stable/generated/torch.norm.html) MPS implementation:

- (A) When executing with `dim=[]` and then executing with `dim=1`, the result is different from the CPU implementation.
- (B) Conversely, when executing with `dim=1` and then executing with `dim=[]`, it causes a crash.

**Code to reproduce issue (A):**

```python
import torch

v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
norm_cpu = v_cpu.norm(2, dim=[])
norm_mps = v_mps.norm(2, dim=[])
print(f"v_mps.norm(2, dim=[]) = {norm_mps} should be {norm_cpu}")

v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
norm_cpu = v_cpu.norm(2, dim=0)
norm_mps = v_mps.norm(2, dim=0)
print(f"v_mps.norm(2, dim=0) = {norm_mps} should be {norm_cpu}")

# prints:
# v_mps.norm(2) = 3.741657257080078 should be 3.7416574954986572
# v_mps.norm(2, 0) = tensor([3.7417, 2.0000, 3.0000], device='mps:0') should be tensor([1., 2., 3.])
```

**Code to reproduce issue (B):**

```python
import torch

v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
norm_cpu = v_cpu.norm(2, dim=0)
norm_mps = v_mps.norm(2, dim=0)
print(f"v_mps.norm(2, dim=0) = {norm_mps} should be {norm_cpu}")

v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
norm_cpu = v_cpu.norm(2, dim=[])
norm_mps = v_mps.norm(2, dim=[])
print(f"v_mps.norm(2, dim=[]) = {norm_mps} should be {norm_cpu}")

# v_mps.norm(2, dim=0) = tensor([1., 2., 3.], device='mps:0') should be tensor([1., 2., 3.])
# /AppleInternal/Library/BuildRoots/c651a45f-806e-11ed-a221-7ef33c48bc85/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayMultiaryKernel.mm:1596: failed assertion `Error: Invalid KernelDAG, equalShape for destination failed'
# zsh: abort      python ./norm_crash.py
```

**I couldn't find a similar issue, so I created this one. ~~I will submit a pull request with a fix later.~~** I've submitted a pull request with a fix https://github.com/pytorch/pytorch/pull/97232.

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0+git37cde56
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.25.2
Libc version: N/A

Python version: 3.9.14 (main, Feb 26 2023, 11:37:37)  [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] mypy==1.0.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] torch==2.1.0a0+gitf561275
[pip3] torchaudio==2.0.0.dev20230224
[pip3] torchvision==0.15.0.dev20230224
[conda] Could not collect
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by ishikawa**
This PR fixes https://github.com/pytorch/pytorch/issues/97231.

## Purpose

The implementation of the `set_axes` function in `aten/src/ATen/native/mps/operations/ReduceOps.mm` is different from the CPU implementation and does not perform an all-reduce when `dim=[]`. As a result, the MPSGraph is cached with incorrect axes.

This leads to the inconsistency and crashes described in https://github.com/pytorch/pytorch/issues/97231.

## Changes made

- Update `set_axes` function in `ReduceOps.mm` to match CPU implementation
- Replace the previous condition for `num_reduce_dims == 0` with a
  check for opt_dim having no value or `num_reduce_dims == 0`
- To perform an all-reduce, modify axes initialization to create an array with `num_input_dims` capacity and set each element to its index value.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 1]
```diff
@@ -40,7 +40,7 @@ void set_apparent_shapes(NSMutableArray<NSNumber*>*& apparent_out_shape,
                          int64_t num_reduce_dims,
                          int64_t num_output_dims,
                          IntArrayRef& input_shape,
-                         NSMutableArray<NSNumber*>*& axes) {
+                         NSArray<NSNumber*>* axes) {
```
Comment (ishikawa) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm:43]: It appears that the set_apparent_shapes function only requires read-only access to the axes argument, so there is no need for it to be an NSMutableArray.

**Comment by kulinseth** [comment]
The diff looks good @ishikawa . Can you please rebase the change?

**Review by kulinseth — APPROVED** [review]

**Comment by kit1980** [comment]
Hi @ishikawa.
I wonder if you're still interested in merging this? If yes, please resolve the conflicts.

**Comment by ishikawa** [comment]
@kulinseth cc: @kit1980 Sorry for too late reply. I've verified the issue is resolved in the `main` branch, so close this PR and issue. Thanks a lot!

```python
>>> import torch
>>> print(torch.__version__)
2.1.0a0+git3f84df3
>>> v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
>>> v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
>>> norm_cpu = v_cpu.norm(2, dim=[])
>>> norm_mps = v_mps.norm(2, dim=[])
>>> print(f"v_mps.norm(2, dim=[]) = {norm_mps} should be {norm_cpu}")
v_mps.norm(2, dim=[]) = 3.7416574954986572 should be 3.7416574954986572
>>> v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
>>> v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
>>> norm_cpu = v_cpu.norm(2, dim=0)
>>> norm_mps = v_mps.norm(2, dim=0)
>>> print(f"v_mps.norm(2, dim=0) = {norm_mps} should be {norm_cpu}")
v_mps.norm(2, dim=0) = tensor([1., 2., 3.], device='mps:0') should be tensor([1., 2., 3.])
```

```python
>>> import torch
>>> print(torch.__version__)
2.1.0a0+git3f84df3
>>> v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
>>> v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
>>> norm_cpu = v_cpu.norm(2, dim=0)
>>> norm_mps = v_mps.norm(2, dim=0)
>>> print(f"v_mps.norm(2, dim=0) = {norm_mps} should be {norm_cpu}")
v_mps.norm(2, dim=0) = tensor([1., 2., 3.], device='mps:0') should be tensor([1., 2., 3.])
>>> v_cpu = torch.tensor([[1.0, 2.0, 3.0]])
>>> v_mps = torch.tensor([[1.0, 2.0, 3.0]], device="mps")
>>> norm_cpu = v_cpu.norm(2, dim=[])
>>> norm_mps = v_mps.norm(2, dim=[])
>>> print(f"v_mps.norm(2, dim=[]) = {norm_mps} should be {norm_cpu}")
v_mps.norm(2, dim=[]) = 3.7416574954986572 should be 3.7416574954986572
```
